### PR TITLE
V4.0 replication (WIP) -> better error handling and error stability

### PR DESCRIPTION
**Pull request by myarichuk**

### CODE REVIEW

**Comment by ayende** [comment]
It is really hard to review a PR with so many merge commits.
If possible, try rebasing your work.

File: src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs [hunk 1]
```diff
@@ -17,36 +15,16 @@ namespace Raven.Server.Documents.Handlers
 {
     public class DocumentReplicationRequestHandler : DatabaseRequestHandler
     {	
-        [RavenAction("/databases/*/documentReplication/changeVector", "GET",
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs:20]: Relation of this change to the commit comment?

File: src/Raven.Server/Documents/Replication/DocumentReplicationTransport.cs [hunk 2]
```diff
@@ -76,10 +76,7 @@ public class DocumentReplicationTransport : IDisposable
         //TODO : add here logic so reconnection is attempted couple of times before giving up
         private async Task<WebSocket> GetAndConnectWebSocketAsync()
         {
-            var uri = new Uri(
-                $@"{_url?.Replace("http://", "ws://")
-                    ?.Replace(".fiddler", "")}/databases/{_targetDbName?.Replace("/", string.Empty)}/documentReplication?srcDbId={_srcDbId}&srcDbName={EscapingHelper
-                        .EscapeLongDataString(_srcDbName)}");
+            var uri = new Uri($"{_url?.Replace("http://", "ws://")?.Replace(".fiddler", "")}/databases/{_targetDbName?.Replace("/", string.Empty)}/documentReplication?srcDbId={_srcDbId}&srcDbName={EscapingHelper.EscapeLongDataString(_srcDbName)}");
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/DocumentReplicationTransport.cs]: Relation of this change to the commit comment?

File: src/Sparrow/Json/BlittableJsonReaderObject.cs [hunk 3]
```diff
@@ -226,7 +226,18 @@ public bool TryGet<T>(string name, out T obj)
             return TryGet(new StringSegment(name,0,name.Length),out obj);
         }
 
-        public bool TryGet<T>(StringSegment name, out T obj)
+	    public bool TryGetDetached(StringSegment name, out BlittableJsonReaderObject obj)
```
Comment (ayende) [inline src/Sparrow/Json/BlittableJsonReaderObject.cs:229]: This is using twice the memory, because it needs to re-read the data.
Why do we need that?

**Comment by ayende** [comment]
Running the fast tests gives me:

```

System.Net.Http.WinHttpException: The connection with the server was terminated abnormally
   at System.Net.WebSockets.WinHttpWebSocket.InternalReceiveAsync(ArraySegment`1 buffer)
   at System.Net.WebSockets.WinHttpWebSocket.<ReceiveAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Raven.Server.Json.WebsocketStream.<ReadAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Sparrow.Json.JsonOperationContext.<ParseToMemoryAsync>d__38.MoveNext()
System.Net.Http.WinHttpException: The connection with the server was terminated abnormally
   at System.Net.WebSockets.WinHttpWebSocket.InternalReceiveAsync(ArraySegment`1 buffer)
   at System.Net.WebSockets.WinHttpWebSocket.<ReceiveAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Raven.Server.Json.WebsocketStream.<ReadAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Sparrow.Json.JsonOperationContext.<ParseToMemoryAsync>d__38.MoveNext()
```

**Comment by ayende** [comment]
Some of those commits are from several weeks ago, there is no need to keep them hidden for so long

**Comment by HibernatingRhinosServiceAccount** [comment]
Looking at the stack trace, it is probably GetLastEtagAsync, which gives us
such beautiful code as:

```
catch (Exception e)
{
return -1;
}
```

That is NOT the kind of code we want.

*Hibernating Rhinos Ltd  *

Oren Eini\* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by ayende** [comment]
Looking at the code, we also have the following warnings:

```
C:\Work\ravendb-4.0\src\Raven.Server\Documents\Replication\DocumentReplicationTransport.cs(77,24): warning CS0168: The variable 'e' is declared but never used
C:\Work\ravendb-4.0\src\Raven.Server\Documents\Replication\DocumentReplicationTransport.cs(176,28): warning CS0168: The variable 'e' is declared but never used
C:\Work\ravendb-4.0\src\Raven.Server\Documents\Handlers\DocumentReplicationHandler.cs(149,25): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
C:\Work\ravendb-4.0\src\Raven.Server\Documents\Replication\OutgoingDocumentReplication.cs(221,46): warning CS0168: The variable 'e' is declared but never used
```

All of which should not be there.

Please fix the aforementioned issues and re-issue a rebased PR.

**Pull request closed without merging**